libraries/lib-snapping/SnapUtils.cpp

```diff
@@ -0,0 +1,315 @@
+   };
+
+   RegistryVisitor registryVisitor { visitor };
+   Registry::TransparentGroupItem<> top { PathStart };
+   Registry::Visit(registryVisitor, &top, &Registry());
+}
+
+SnapRegistryItem* SnapFunctionsRegistry::Find(const Identifier& id)
+{
+   using Cache = std::unordered_map<Identifier, SnapRegistryItem*>;
+   static Cache cache;
+
+   auto it = cache.find(id);
+   if (it != cache.end())
+      return it->second;
+
+   struct CacheUpdater final : Registry::Visitor
+   {
+      explicit CacheUpdater(Cache& _cache)
+          : cache(_cache)
+      {
+      }
+
+      void Visit(Registry::SingleItem& item, const Path&) override
+      {
+         auto concreteItem = dynamic_cast<SnapRegistryItem*>(&item);
+
+         if (concreteItem == nullptr)
+            return;
+
+         auto it = cache.find(concreteItem->name);
+
+         if (it == cache.end())
+         {
+            cache.insert(
+               std::make_pair(concreteItem->name, concreteItem));
+         }
+      }
+
+      Cache& cache;
+   };
+
+   CacheUpdater update { cache };
+   Registry::Visit(update, &Registry());
+
+   it = cache.find(id);
+
+   return it != cache.end() ? it->second : nullptr;
+}
+
+SnapResult SnapFunctionsRegistry::Snap(
+   const Identifier& id, SnapConfig config, double time, bool nearest)
+{
+   auto item = Find(id);
+
+   if (item == nullptr)
+      return SnapResult { time, false };
+
+   return item->snapFunction(config, time, nearest);
+}
+
+SnapRegistryGroup::~SnapRegistryGroup()
+{
+}
+
+bool SnapRegistryGroup::Transparent() const
+{
+   return transparent;
+}
+
+SnapRegistryItem::SnapRegistryItem(
+   const Identifier& internalName, const TranslatableString& _label,
+   SnapFunctor _snapFunction)
+    : SingleItem { internalName }
+    , label { _label }
+    , snapFunction { _snapFunction }
+{
+}
+
+SnapRegistryItem::~SnapRegistryItem()
+{
+}
+
+Registry::BaseItemPtr SnapFunction(
+   const Identifier& functionId, const TranslatableString& label,
+   SnapFunctor functor)
+{
+   return std::make_unique<SnapRegistryItem>(
+      functionId, label, std::move(functor));
+}
+
+SnapRegistryItemRegistrator::SnapRegistryItemRegistrator(
+   const Registry::Placement& placement, Registry::BaseItemPtr pItem)
+    : RegisteredItem { std::move(pItem), placement }
+{
+}
+
+namespace
+{
+SnapResult SnapWithMultiplier (double value, double multiplier, bool nearest)
+{
+   auto result = nearest ? std::round(value * multiplier) / multiplier :
+                           std::floor(value * multiplier) / multiplier;
+
+   return SnapResult { result, true };
+}
+
+
+SnapFunctor SnapToTime(double multiplier) {
+   return [multiplier](SnapConfig, double value, bool nearest)
+   { return SnapWithMultiplier(value, multiplier, nearest); };
+}
+
+SnapResult SnapToSamples (SnapConfig cfg, double value, bool nearest)
+{
+   return SnapWithMultiplier(value, cfg.rate, nearest);
+}
+
+SnapFunctor SnapToFrames(double fps)
```

Comment:
This is just like `SnapToTime`.  Will it be changed?

Reply:
No. I like it to have an explicit name.